src/ucp/core/ucp_ep.c

```diff
@@ -3253,24 +3254,78 @@ ucp_ep_is_am_keepalive(ucp_ep_h ep, ucp_rsc_index_t rsc_idx)
     }
 
     wiface = ucp_worker_iface(ep->worker, rsc_idx);
-    return ucs_test_all_flags(wiface->attr.cap.flags,
-                              UCT_IFACE_FLAG_CONNECT_TO_IFACE |
-                              UCT_IFACE_FLAG_AM_BCOPY);
+    return (lane != UCP_NULL_LANE) &&
+           !(ucp_ep_config(ep)->p2p_lanes & UCS_BIT(lane)) &&
+           (wiface->attr.cap.flags & UCT_IFACE_FLAG_AM_BCOPY);
 }
 
-ucs_status_t ucp_ep_do_uct_ep_keepalive(ucp_ep_h ucp_ep, uct_ep_h uct_ep,
-                                        ucp_rsc_index_t rsc_idx, unsigned flags,
+static UCS_F_ALWAYS_INLINE uct_ep_h
+ucp_ep_select_keepalive(ucp_ep_h ucp_ep, ucp_rsc_index_t *rsc_idx_p,
+                        ucp_lane_index_t *lane_p)
+{
+    ucp_lane_index_t lane      = ucp_ep_config(ucp_ep)->key.keepalive_lane;
+    uct_ep_h uct_ep            = ucp_ep->uct_eps[lane];
+    ucp_wireup_ep_t *wireup_ep = ucp_wireup_ep(uct_ep);
+    ucp_rsc_index_t rsc_idx;
+
+    if (wireup_ep != NULL) {
```

Comment:
i think it looks weird that we use this function twice, once from wireup ep and another time from do_keepalive, and then have a branch here to handle each case differently.
also not clear why need this refactoring, shouldn't we change only ucp_ep_is_am_keepalive() to fix the logic?

Reply:
I fixed the `ucp_ep_is_am_keepalive`, but I was asked to not pass `lane` which requires to move the uct_ep/rsc_index initialization logic to `ucp_ep_do_uct_ep_keepalive`

Reply:
https://github.com/openucx/ucx/pull/8113/commits/b26180772cec01e8699f4ad2291b1208046be2ce - original fix